crates/stages/src/trie/mod.rs

```diff
@@ -0,0 +1,467 @@
+            ChildReference::Hash(ref hash) => hash.as_ref(),
+            ChildReference::Inline(ref _inline_data, _len) => {
+                unreachable!("can't happen")
+            }
+        };
+
+        let mut output = Vec::new();
+        encode_iter([encoded_partial, value].into_iter(), &mut output);
+        output
+    }
+
+    fn branch_node(
+        children: impl Iterator<Item = impl Borrow<Option<ChildReference<Self::HashOut>>>>,
+        maybe_value: Option<Value<'_>>,
+    ) -> Vec<u8> {
+        let mut output = Vec::new();
+        let mut children: Vec<_> = children
+            .map(|c| -> Vec<u8> {
+                match c.borrow() {
+                    Some(ChildReference::Hash(hash)) => hash.as_ref().to_vec(),
+                    Some(ChildReference::Inline(_value, _len)) => {
+                        unreachable!("can't happen because all keys are equal length");
+                    }
+                    None => vec![],
+                }
+            })
+            .collect();
+
+        children.push(match maybe_value {
+            Some(Value::Inline(value)) => value.to_vec(),
+            None => vec![],
+            _ => unimplemented!("unsupported"),
+        });
+        encode_iter(children.iter().map(|c| c.as_slice()), &mut output);
+        output
+    }
+
+    fn branch_node_nibbled(
+        _partial: impl Iterator<Item = u8>,
+        _number_nibble: usize,
+        _children: impl Iterator<Item = impl Borrow<Option<ChildReference<<H as Hasher>::Out>>>>,
+        _value: Option<Value<'_>>,
+    ) -> Vec<u8> {
+        unimplemented!("unused");
+    }
+}
+
+/// An Ethereum account.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Default, RlpEncodable, RlpDecodable)]
+struct EthAccount {
+    /// Account nonce.
+    nonce: u64,
+    /// Account balance.
+    balance: U256,
+    /// Account's storage root.
+    storage_root: H256,
+    /// Hash of the account's bytecode.
+    code_hash: H256,
+}
+
+impl From<Account> for EthAccount {
+    fn from(acc: Account) -> Self {
+        EthAccount {
+            nonce: acc.nonce,
+            balance: acc.balance,
+            storage_root: EMPTY_ROOT,
+            code_hash: acc.bytecode_hash.unwrap_or(KECCAK_EMPTY),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct DBTrieLoader;
+
+impl DBTrieLoader {
+    pub(crate) fn calculate_root<DB: Database>(
+        &mut self,
+        tx: &Transaction<'_, DB>,
+    ) -> Result<H256, TrieError> {
+        let mut accounts_cursor = tx.cursor_read::<tables::HashedAccount>()?;
+        let mut walker = accounts_cursor.walk(H256::zero())?;
+
+        // TODO: implement HashDB to use DB instead of MemoryDB
+        // let trie_cursor = tx.cursor_read::<tables::AccountsTrie>()?;
+
+        let mut db = MemoryDB::<KeccakHasher, HashKey<KeccakHasher>, Vec<u8>>::from_null_node(
+            RLPNodeCodec::<KeccakHasher>::empty_node(),
+            RLPNodeCodec::<KeccakHasher>::empty_node().to_vec(),
+        );
```

Comment:
Is this going to save everything in RAM as it will not be usable for latest state?

Reply:
No. This was only temporary until I finished adding the bindings between the trie and our database.